src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+					md->omd_ver);
+				spare_avail = false;
+				goto next_fail;
+			}
+
+			/*
+			* The selected spare is down prior to current failed
+			* one, then it can't be a valid spare, let's skip it
+			* and try next spare.
+			*/
+			if (spare_tgt->ta_comp.co_fseq < f_shard->fs_fseq)
+				continue; /* try next spare */
+
+			/*
+			* If both failed target and spare target are down, then
+			* add the spare target to the fail list for remap, and
+			* try next spare..
+			*/
+			if (f_shard->fs_status == PO_COMP_ST_DOWN)
+				D_ASSERTF(spare_tgt->ta_comp.co_status !=
+					  PO_COMP_ST_DOWNOUT,
+					  "down fseq(%u) < downout fseq(%u)\n",
+					  f_shard->fs_fseq,
+					  spare_tgt->ta_comp.co_fseq);
+
+			f_shard->fs_fseq = spare_tgt->ta_comp.co_fseq;
+			f_shard->fs_status = spare_tgt->ta_comp.co_status;
+
+			current = current->next;
+			d_list_del_init(&f_shard->fs_list);
+			remap_add_one(remap_list, f_shard);
+
+			/* Continue with the failed shard has minimal fseq */
+			if (current == remap_list) {
+				current = &f_shard->fs_list;
+			} else {
+				f_tmp = d_list_entry(current,
+						     struct failed_shard,
+						     fs_list);
+				if (f_shard->fs_fseq < f_tmp->fs_fseq)
+					current = &f_shard->fs_list;
+			}
+			continue; /* try next spare */
+		}
+next_fail:
+		fail_count = 0;
+		if (spare_avail) {
+			/* The selected spare target is up and ready */
+			l_shard->po_target = spare_tgt->ta_comp.co_id;
+			l_shard->po_fseq = f_shard->fs_fseq;
+
+			/*
+			* Mark the shard as 'rebuilding' so that read will
+			* skip this shard.
+			*/
+			if (f_shard->fs_status == PO_COMP_ST_DOWN) {
+				l_shard->po_rebuilding = 1;
+				f_shard->fs_tgt_id = spare_tgt->ta_comp.co_id;
+			}
+		} else {
+			l_shard->po_shard = -1;
+			l_shard->po_target = -1;
+		}
+		current = current->next;
+	}
+
+	mapless_remap_dump(remap_list, md, "after remap:");
+	return 0;
+}
+
+static int
+mapless_obj_spec_place_get(struct pl_mapless_map *mmap, daos_obj_id_t oid,
+			   struct pool_target **target, uint8_t *dom_used,
+			   uint32_t dom_bytes)
+{
+	struct pool_target      *tgts;
+	unsigned int            tgts_nr;
+	d_rank_t                rank;
+	int                     tgt;
+	unsigned int            pos;
+	int rc;
+
+	D_ASSERT(daos_obj_id2class(oid) == DAOS_OC_R3S_SPEC_RANK ||
+		 daos_obj_id2class(oid) == DAOS_OC_R1S_SPEC_RANK ||
+		 daos_obj_id2class(oid) == DAOS_OC_R2S_SPEC_RANK);
+
+	/* locate rank in the pool map targets */
+	tgts = pool_map_targets(mmap->mmp_map.pl_poolmap);
+	tgts_nr = pool_map_target_nr(mmap->mmp_map.pl_poolmap);
+	/* locate rank in the pool map targets */
+	rank = daos_oclass_sr_get_rank(oid);
+	tgt = daos_oclass_st_get_tgt(oid);
+
+	for (pos = 0; pos < tgts_nr; pos++) {
+		if (rank == tgts[pos].ta_comp.co_rank &&
+		    (tgt == tgts[pos].ta_comp.co_index))
+			break;
+	}
+	if (pos == tgts_nr)
+		return -DER_INVAL;
+
+	*target = &(tgts[pos]);
+
+
+	struct pool_domain *current_dom;
+	struct pool_domain *root;
```

Comment:
It looks violating the coding style, should be put at the beginning of the function.